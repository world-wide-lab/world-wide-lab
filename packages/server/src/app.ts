import path from "path";
import cors from "cors";
import express, { Request, Response } from "express";
import helmet from "helmet";

import { admin, adminRouter } from "./admin";
import api from "./api";
import apiDocs from "./api-docs";
import { routerProtectedWithoutAuthentication } from "./api/protected";
import config from "./config";
import { logger } from "./logger";

const app = express();

// app.use(helmet());
app.use(cors());
app.use(express.json());

// Log all HTTP requests
if (config.logging.http) {
  logger.info("Logging HTTP requests");

  app.use((req, res, next) => {
    logger.http(`${req.method} ${req.url} ${JSON.stringify(req.body)}`, {
      request: {
        query: req.query,
        params: req.params,
        body: req.body,
        headers: req.headers,
      },
    });
    next();
  });
}

app.use("/static", express.static(path.join(__dirname, "..", "static")));

app.get("/", async (req: Request, res: Response) => {
  res.send(`
  <!DOCTYPE html>
  <html lang="en">
    <head>
      <meta charset="utf-8">
      <title>World-Wide-Lab</title>
    </head>
    <body>
      World-Wide-Lab is running 🌐🧑‍🔬👩‍🔬👨‍🔬 <br />
      <ul>
        <li>The latest API can be found under <a href="/v1/">/v1/</a></li>
      </ul>
    </body>
  </html>
  `);
});

// Mount the API router under v1
app.use("/v1", api);

// Mount the api-docs
if (config.apiDocs.enabled) {
  app.use("/api-docs", apiDocs);
}

// Use adminJS
if (config.admin.enabled) {
  // Make protected API routes available in Admin UI
  adminRouter.use("/wwl/", routerProtectedWithoutAuthentication);

  app.use(admin.options.rootPath, adminRouter);
}

export default app;
