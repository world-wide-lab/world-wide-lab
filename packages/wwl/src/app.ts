import path from 'path'
import express, {Request, Response} from 'express';
import helmet from 'helmet';
import cors from 'cors';

import api from './api';
import apiDocs from './api-docs';
import { admin, adminRouter } from './admin'
import config from './config'

const app = express();

// app.use(helmet());
app.use(cors());
app.use(express.json());

app.use('/static', express.static(path.join(__dirname, '..', 'static')))

app.get("/", async (req: Request, res: Response) => {
  res.redirect("/v1/");
});

// Mount the API router under v1
app.use('/v1', api);

// Mount the api-docs
if (config.apiDocs.enabled) {
  app.use('/api-docs', apiDocs);
}

// Use adminJS
if (config.admin.enabled) {
  app.use(admin.options.rootPath, adminRouter)
}

export default app;
