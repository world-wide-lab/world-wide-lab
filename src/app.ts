import 'dotenv/config';
import express, {Request, Response} from 'express';
import helmet from 'helmet';
import cors from 'cors';

import api from './api';

const app = express();

app.use(helmet());
app.use(cors());
app.use(express.json());

app.get("/", async (req: Request, res: Response) => {
  res.redirect("/v1/");
});

// Mount the API router under v1
app.use('/v1', api);

export default app;
