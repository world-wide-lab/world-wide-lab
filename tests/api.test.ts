// Set up fake environment variables
import "./setup_env";

import request from 'supertest';
import sequelize from '../src/db';
import app from '../src/app';

const STUDY_ID = 'abc123';

describe('API Routes', () => {
  beforeAll(async () => {
    // Initialize Database
    await sequelize.sync();
  });

  // afterAll(async () => {
  //   // TODO: Clean up any test data
  // });

  let participantId: string;
  let runId: string;
  let studyId: string = STUDY_ID;

  describe('POST /participant', () => {
    it('should create a new participant', async () => {
      const response = await request(app)
        .post('/v1/participant')
        .send();

      expect(response.status).toBe(200);
      expect(response.body).toHaveProperty('participantId');

      // Set the shared participantId
      participantId = response.body.participantId;
    });
  });

  describe('PUT /participant/:participantId', () => {
    it('should update an existing participant', async () => {
      const response = await request(app)
        .put('/v1/participant/' + participantId)
        .send({ info: { lorem: 'ipsum' } });

      expect(response.status).toBe(200);
      expect(response.body).toMatchSnapshot();

      const user = await sequelize.models.Participant.findOne({
        where: { participantId }
      })
      expect(user).toHaveProperty('info', { lorem: 'ipsum' });
    });
  });

  describe('POST /study', () => {
    it('should create a new study', async () => {
      const response = await request(app)
        .post('/v1/study')
        .send({ studyId: STUDY_ID});

      expect(response.status).toBe(200);
      expect(response.body).toHaveProperty('studyId');
    });
  });

  describe('POST /run', () => {
    it('should start a new run', async () => {
      const response = await request(app)
        .post('/v1/run')
        .send({ participantId, studyId });

      expect(response.status).toBe(200);
      expect(response.body).toHaveProperty('participantId', participantId);
      expect(response.body).toHaveProperty('studyId', studyId);

      // Set the shared runId
      runId = response.body.runId;
    });
  });

  describe('POST /run/finish', () => {
    it('should mark a run as finished', async () => {
      const response = await request(app)
        .post('/v1/run/finish')
        .send({ runId: runId });

      expect(response.status).toBe(200);
      expect(response.body).toMatchSnapshot();

      const run = await sequelize.models.Run.findOne({
        where: { runId }
      })
      expect(run).toHaveProperty('finished', true);
    });
  });

  describe('PUT /run/:runId', () => {
    it('should update a run', async () => {
      const response = await request(app)
        .put('/v1/run/' + runId)
        .send({ info: { lorem: 'ipsum' } });

      expect(response.status).toBe(200);
      expect(response.body).toMatchSnapshot();

      const run = await sequelize.models.Run.findOne({
        where: { runId }
      })
      expect(run).toHaveProperty('info', { lorem: 'ipsum' });
    });
  });

  describe('POST /response', () => {
    it('should submit a response', async () => {
      const response = await request(app)
        .post('/v1/response')
        .send({
          runId,
          name: 'test_trail',
          payload: {
            key_1: 'value 1',
            key_2: 'value 2',
          },
        });

      expect(response.status).toBe(200);
      expect(response.body).toHaveProperty('responseId');
      const { responseId } = response.body;

      const runResponse = await sequelize.models.Response.findOne({
        where: { responseId }
      })
      expect(runResponse).toHaveProperty('runId', runId);
      expect(runResponse).toHaveProperty('name', 'test_trail');
      expect(runResponse).toHaveProperty('payload', { key_1: 'value 1', key_2: 'value 2', });
    });
  });
});
