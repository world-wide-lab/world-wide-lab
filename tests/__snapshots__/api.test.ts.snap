// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`API Routes POST /run/finish should mark a run as finished 1`] = `
[
  1,
]
`;

exports[`API Routes PUT /participant/:participantId should update an existing participant 1`] = `
[
  1,
]
`;

exports[`API Routes PUT /run/:runId should update a run 1`] = `
[
  1,
]
`;
